File: _java_/src/main/java/examples/algorithms/LRUCache.java
package examples.algorithms;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Least recently used cache.
 * If size is exceeded, the least recently used item is removed.
 */
class LRUCache {
    private final Map<Integer, Integer> cache;
    private final int capacity;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        // true sets the map to access order mode,
        // which makes recently accessed items (get or put) appear at the end of the list.
        this.cache = new LinkedHashMap<>(capacity, 0.75f, true);
    }

    public int get(int key) {
        return cache.getOrDefault(key, -1);
    }

    public void put(int key, int value) {
        cache.put(key, value);

        if (cache.size() > capacity) {
            // Remove the least recently used entry (the first entry in the map)
            Map.Entry<Integer, Integer> firstEntry = cache.entrySet().iterator().next();
            cache.remove(firstEntry.getKey());
        }
    }

    public String toString() {
        return cache.toString();
    }

    public static void main(String[] args) {
        LRUCache obj = new LRUCache(2);
        obj.get(2);
        System.out.println(obj);
        obj.put(2, 6);
        System.out.println(obj);
        obj.get(1);
        System.out.println(obj);
        obj.put(1, 5);
        System.out.println(obj);
        obj.put(1, 2);
        System.out.println(obj);
        obj.get(1);
        System.out.println(obj);
        obj.get(2);
        System.out.println(obj);
    }
}

File: _python_/algorithms/lru_cache.py
from collections import OrderedDict


class LRUCache:
    """
    Least recently used cache.
    If size is exceeded, the least recently used item is removed.
    """

    def __init__(self, capacity: int):
        self._cache = OrderedDict()
        self._capacity = capacity

    def get(self, key: int) -> int:
        if key in self._cache:
            self._cache.move_to_end(key)  # Mark as recently used
            return self._cache[key]
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self._cache:
            self._cache.move_to_end(key)

        self._cache[key] = value

        if len(self._cache) > self._capacity:
            self._cache.popitem(last=False)  # Remove least recently used item

    def __repr__(self):
        return str(self._cache)


lru_cache = LRUCache(2)
lru_cache.put(1, 1)
lru_cache.put(2, 2)
print(lru_cache)

lru_cache.get(1)
print(lru_cache)

lru_cache.put(3, 3)
print(lru_cache)

lru_cache.get(2)
lru_cache.put(4, 4)
print(lru_cache)

lru_cache.get(1)
lru_cache.get(3)
lru_cache.get(4)

print(lru_cache)
